Add code displaying and copying feature, solve icon exceptions

tarvos21 opened the pull request:
- generate code and display
- put JavaScript after each module
- make sidebars work properly
- style definitions

tarvos21 committed:
Dependencies for docs.js

- clipboard.js: copy code to system clipboard
- cookie.js: prevent code icon from popuping "View Source" hint after first
  visit
- highlight.min.js: highligh code, not working properly at present
- less.js: parse html, not quite clear about what it does

tarvos21 committed:
Move JavaScript code of each module from kitchen-sink.js to index.slim

- move code from kitchen-sink.js to correspinding modules
- remain kitchen-sink.js for now, it could be deleted
- add the line left to default.js

tarvos21 committed:
Make selector of the four main tabs more specific, not influencing tab example

tarvos21 committed:
Make sidebars work properly

- Move sidebars in four directions to sidebar.slim
- Make sidebars direct childrens of body
- Put main content in a div with class "pusher"

Problems:
- performance is not stable, sometimes it works incorrectly
- refresh with F5 may help

tarvos21 committed:
Make "Example" label segment work consitently

- Headers like "Button", "Transition" appear outside the example segment
- Make examples "Image" and "Dropdown" display normally as other examples

tarvos21 committed:
Style settings for code displaying

- Copy icon, code icon, example label bar, code area

Others:
- make grids in grid example in same width
- change z-index of piled segment's parent

tarvos21 committed:
Create HTML/JavaScript code and display

- generate, create HTML, then format

Problems remaining:
- no highlight: highlightjs.org
- several lines contains two many image/i tags, not great flaw, maybe good

xiaohanyu commented on layouts/footer.slim:
```diff
@@ -5,5 +5,9 @@ footer.ui.center.aligned.container lang='en'
 script src="/static/javascripts/jquery.address.js"
+script src="/static/javascripts/cookie.js"
+script src="/static/javascripts/clipboard.js"
+script src="/static/javascripts/less.js"
```
To answer your question, "- less.js: parse html, not quite clear about what it does", it seems that by loading a `less.js`, you can write `less` code directly in a web page.

tarvos21 replied:
Great, so this `less.js` is not necessary now?

xiaohanyu replied:
Not quite sure about this. I've tested by remove this line and the code clipboard don't work anymore.

So keep it here, I think it's OK right now.

xiaohanyu commented on layouts/default.slim:
```diff
@@ -3,9 +3,13 @@ html
-    #content
-      == render '/menu.*'
-      == yield
+    .pusher
```
I think it's better to create a new layout and nanoc rules for kitchen-sink page, since not all pages need a sidebar. It's OK for this PR, I'll do the later work to refactoring layout.

tarvos21 replied:
Ok, it can be done somehow. I do it because the semanic.min.js keeps telling me that it's better to make the sidebar as the direct children of body tag. But it's not a great problem.

xiaohanyu replied:
yep, agree.

xiaohanyu commented:
https://github.com/semantic-ui-forest/semantic-ui-forest/pull/5/commits/4ee62d887fa0d053494d897e3a723e93da7378ee

Typo in commit message: "Iamge" -> "Image"

xiaohanyu commented on content/partials/examples/dropdown.slim:
```diff
@@ -798,7 +798,7 @@
       |  Tag D
 .ui.hidden.divider
-|  Show me posts trending
+span Show me posts trending 
```
Not sure why you add a `span` here. 

And second, if possible, better to remove the extra space in the end of this line.

tarvos21 replied:
It's an interesting question. Let me explain it here: in the original version, this string "Show me posts trending" is naked in the ".example" div, so if you click the code icon to generate code and example segment, the string without any class or tag will be filtered out from the example segment, it will appear right down the header "Dropdown" without any decoration. So is the problem with the string "User Avatar" in image example. 

As to the whitespace, it is added for the sentence "Show me posts trending **today**", "today" is another single bold div with a dropdown, if no whitespace added here, "trending" will be right next to "today": "Show me posts trending**today**"

Oh, the problem can still be seen here:
http://semantic-ui.com/kitchen-sink.html#dropdown
http://semantic-ui.com/kitchen-sink.html#image
The string never goes back even if you make the code disappear.


Pictures may be easier to demonstrate:

**Dropdown**:
![23-5](https://cloud.githubusercontent.com/assets/18745870/21453923/0f55a176-c94f-11e6-90e1-018db4b79a77.png)
![23-6](https://cloud.githubusercontent.com/assets/18745870/21453927/19ea63e2-c94f-11e6-873a-548af1ed2c44.png)

**Image**:
![23-7](https://cloud.githubusercontent.com/assets/18745870/21453946/425471ec-c94f-11e6-8ccd-83e0415f398f.png)
![23-8](https://cloud.githubusercontent.com/assets/18745870/21453948/462ea544-c94f-11e6-9ebb-dbcf5ecb913e.png)

xiaohanyu replied:
Cool, thanks for your detailed explanation.

xiaohanyu commented on content/partials/examples/dimmer.slim:
```diff
@@ -1,63 +1,64 @@
-    .ui.dimmer
-  .ui.ignored.ignored.icon.buttons
-    .ui.show.button
+.ui
```
Not sure why you add a single `.ui` element here. I've never seen a single `.ui` in semantic-ui. Since in semantic-ui, `.ui` is always combined with other class like `.button`, `.menu`, it's just a tag to show that the element is a UI element.

tarvos21 replied:
Also related to the elements filter in the javascript. Without ".ui", the filter will work inconsitently.

http://semantic-ui.com/kitchen-sink.html#/elements
http://semantic-ui.com/kitchen-sink.html#/collections

If you click the code icon in **Button** and **Breadcrumb** example on those two page, you can see  that in the Button example, the header "Button" is **inside** the example segment, while in the BreadCrumb example, the header "Breadcrumb" is **outside** the example segment. 
This is the main reason to do such modifications.

![23-9](https://cloud.githubusercontent.com/assets/18745870/21453999/bf77de48-c94f-11e6-88e1-b0533e6f2265.png)
![23-10](https://cloud.githubusercontent.com/assets/18745870/21454000/c19c05b4-c94f-11e6-88ec-20163c0b0ad6.png)

tarvos21 replied:
Another thing to supplement, the reason I use single ".ui" somewhere is that, if I use ".ui.button.demo" instead of ".ui.demo", the whole area will become a huge button, other parts also have the problem, like rating, sidebar. ".demo" is not necessary in some parts, just leave it now.

xiaohanyu replied:
OK, thanks!

xiaohanyu reviewed (changes requested):
Wonderful work. 

Only found a tiny typo. Leave two questions here.

tarvos21 commented:
Another thing can be done is: create single module.js files and put the modules' js code to those files, then  render them through Ruby. It's another possible solution if you feel the index.slim file is too long.

tarvos21 committed:
Change navbar dropdown icon on GitHub theme

- change from '\f0da' to '\f05a'
- a little bigger than other themes

tarvos21 committed:
Add icon font of Material theme

- Add icon font from Semantic-UI repository
- Let code icon be '\f121', same with other themes(optional)

Others:
- copy icon in "Example" label segment varies in GitHub, Material themes

xiaohanyu reviewed (approved):
Wonderful work, thanks thanks!

tarvos21 commented:
Very happy to hear what you say!
